lib/skate/detours/route_segments.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Skate.Detours.RouteSegments do
+  @moduledoc """
+  Break a route into segments based on `onnection_start` and `connection_end` points
+  """
+
+  defmodule Params do
+    @moduledoc false
+    @type t :: %__MODULE__{
+            connection_start: Util.Location.From.t(),
+            connection_end: Util.Location.From.t(),
+            shape: nonempty_list(Util.Location.From.t())
+          }
+    @enforce_keys [:connection_start, :connection_end, :shape]
+    defstruct [:connection_start, :connection_end, :shape]
+  end
```

Comment:
**suggestion:** given that all three of these are required, I think it would be more idiomatic to just define `route_segments` as a function that takes 3 arguments rather than 1 and not define an entirely separate `Params` struct.

Reply:
I think the way I did it is the only way to make the lexical order of `shape`, `connection_start`, and `connection_end` not matter. If I used a keyword list, then they have to be in the right order (which makes keyword lists super annoying to work with! 😠), and with just a regular arg list, they still have to be in the right order and it's even harder to remember.

I don't feel that strongly about it though, so happy to do whatever makes sense.

Reply:
There are plenty of arity-3 functions so I don't think it's particularly hard to remember (I'd feel differently if we were talking 5+ arguments or something). Also, thanks to `@doc` and `@spec` you can get good inline documentation in your editor via LSP when you go to use the function. Plus, using a struct rather than individual arguments makes it harder to use in pipelines.

Reply:
https://github.com/mbta/skate/pull/2438/commits/434618f81afa36e2eaff1bafade928165ffd6e88